### PR TITLE
Support status argument for GatherXXX

### LINKED ISSUE #4466: Variant of `GatherAlpha()` that takes a `status` argument is not supported
This is likely low priority, so filing mostly for documentation purposes right now. 

Currently, Slang does not support the variant of `GatherAlpha()` that takes a status argument:
```
shader.hlsl(11): error 39999: no overload for 'GatherAlpha' applicable to arguments of type (SamplerState, vector<float,3>, vector<int,2>, uint)
    uint4 e = t2u4.GatherAlpha(gSampler, location, int2(1, 2), status);
                              ^
hlsl.meta.slang(2329): note 39999: candidate: func __TextureImpl<vector<T,N>, Shape, isArray, 0, sampleCount, 0, isShadow, 0, format>.GatherAlpha(SamplerState, float3, int2, int2, int2, int2) -> uint4
hlsl.meta.slang(2315): note 39999: candidate: func __TextureImpl<vector<T,N>, Shape, isArray, 0, sampleCount, 0, isShadow, 0, format>.GatherAlpha(SamplerState, float3, int2) -> uint4
hlsl.meta.slang(2301): note 39999: candidate: func __TextureImpl<vector<T,N>, Shape, isArray, 0, sampleCount, 0, isShadow, 0, format>.GatherAlpha(SamplerState, float3) -> uint4
shader.hlsl(12): error 39999: no overload for 'GatherAlpha' applicable to arguments of type (SamplerState, vector<float,3>, vector<int,2>, vector<int,2>, vector<int,2>, vector<int,2>, uint)
    uint4 f = t2u4.GatherAlpha(gSampler, location, int2(1, 2), int2(3, 4), int2(5, 6), int2(7, 8), status);
                              ^
hlsl.meta.slang(2329): note 39999: candidate: func __TextureImpl<vector<T,N>, Shape, isArray, 0, sampleCount, 0, isShadow, 0, format>.GatherAlpha(SamplerState, float3, int2, int2, int2, int2) -> uint4
hlsl.meta.slang(2315): note 39999: candidate: func __TextureImpl<vector<T,N>, Shape, isArray, 0, sampleCount, 0, isShadow, 0, format>.GatherAlpha(SamplerState, float3, int2) -> uint4
hlsl.meta.slang(2301): note 39999: candidate: func __TextureImpl<vector<T,N>, Shape, isArray, 0, sampleCount, 0, isShadow, 0, format>.GatherAlpha(SamplerState, float3) -> uint4
```

**CMD**
```
slangc.exe -target spirv -lang slang -D__spirv__ -emit-spirv-directly -profile ps_6_0 -entry main shader.hlsl
```

**Shader**
```
SamplerState gSampler : register(s5);

Texture2DArray<float4>  t2f4       : register(t1);
Texture2DArray<uint4>   t2u4       : register(t2);
TextureCubeArray<uint4> tCubeArray : register(t3);

float4 main(float3 location: A) : SV_Target {
    uint status;
    uint4 e = t2u4.GatherAlpha(gSampler, location, int2(1, 2), status);
    uint4 f = t2u4.GatherAlpha(gSampler, location, int2(1, 2), int2(3, 4), int2(5, 6), int2(7, 8), status);
    uint4 g = tCubeArray.GatherAlpha(gSampler, float4(1.5, 1.5, 1.5, 1.5), status);
    return (float4)(e + f + g);
}
```

### PR DESCRIPTION
**Pull request by jkwak-work**
This commit adds an argument to all texture GatherXXX functions. The new argument is for "status" as described in SM5.0 definision.

Close #4466

### CODE REVIEW

File: source/slang/hlsl.meta.slang [hunk 1]
```diff
@@ -2296,10 +2296,14 @@ ${{{{
         {
             auto componentName = componentNames[componentId];
             auto glslComponent = (isCmp ? "" :glslComponentNames[componentId == 0 ? 0 : componentId - 1]);
-    }}}}
+    
+            for (bool isStatus : { false, true })
+            {
+                const char* statusDecl = isStatus ? ", uint status" : "";
```
Comment (ArielG-NV) [inline source/slang/hlsl.meta.slang:2302]: status is an `out` [parameter](https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/t2d-gather-s-float-int-uint-)
```suggestion
                const char* statusDecl = isStatus ? ", out uint status" : "";
```

File: source/slang/hlsl.meta.slang [hunk 2]
```diff
@@ -2341,10 +2346,11 @@ ${{{{
             return __texture_gather$(cmp)_offsets<T>(this,$(samplerStateParam) location $(compareArg), offset1,offset2,offset3,offset4 $(glslComponent));
         }
     }
-    ${{{{
+${{{{
+            } // for (isStatus)
```
Comment (ArielG-NV) [inline source/slang/hlsl.meta.slang:2350]: (not a suggestion)
I really like the idea of doing this to make it easier to read inline C++, I will be doing this from now on as well.

**Comment by ArielG-NV** [comment]
I wonder if we should add a stub for [CheckAccessFullyMapped()](https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/checkaccessfullymapped)?

File: source/slang/hlsl.meta.slang [hunk 3]
```diff
@@ -2313,7 +2317,7 @@ ${{{{
     }
     [ForceInline]
     [require(glsl_hlsl_metal_spirv, texture_gather)]
-    vector<T,4> Gather$(cmp)$(componentName)($(samplerStateType)$(samplerStateParam) vector<float, Shape.dimensions+isArray> location $(cmpParam), constexpr vector<int, Shape.planeDimensions> offset)
+    vector<T,4> Gather$(cmp)$(componentName)($(samplerStateType)$(samplerStateParam) vector<float, Shape.dimensions+isArray> location $(cmpParam), constexpr vector<int, Shape.planeDimensions> offset $(statusDecl))
```
Comment (csyonghe) [inline source/slang/hlsl.meta.slang:2320]: We need to initialize status to 0 for all other targets.

File: source/slang/hlsl.meta.slang [hunk 4]
```diff
@@ -2296,11 +2296,17 @@ ${{{{
...
+            {
+                const char* statusDecl = isStatus ? ", out uint status" : "";
+                const char* statusInit = isStatus ? " status = 0;\n" : "";
+}}}}
     [ForceInline]
     [require(glsl_hlsl_metal_spirv, texture_gather)]
-    vector<T,4> Gather$(cmp)$(componentName)($(samplerStateType)$(samplerStateParam) vector<float, Shape.dimensions+isArray> location $(cmpParam))
+    vector<T,4> Gather$(cmp)$(componentName)($(samplerStateType)$(samplerStateParam) vector<float, Shape.dimensions+isArray> location $(cmpParam) $(statusDecl))
     {
+        $(statusInit)
```
Comment (ArielG-NV) [inline source/slang/hlsl.meta.slang:2309]: I was looking into what `status` means, and it seems that this is to signify if "sparse memory accessed is resident"

The current implementations [needs to be changed](https://[registry.khronos.org/OpenGL/extensions/ARB/ARB_sparse_texture2.txt) to `sparseImage*` operations

**Comment by ArielG-NV** [comment]
> I wonder if we should add a stub for [CheckAccessFullyMapped()](https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/checkaccessfullymapped) so code that uses `status` can compile (`status` is useless without `CheckAccessFullyMapped()`, so these functions appear in pairs)?

After looking into what `status` means across platforms, `CheckAccessFullyMapped()` must be added for sparse accesses to work properly.

HLSL: `CheckAccessFullyMapped()`
GLSL: `sparseTexelsResidentARB()` with extension `GL_ARB_sparse_texture2`
SPIRV: `OpImageSparseTexelsResident` with capability `OpCapability SparseResidency`

**Comment by jkwak-work** [comment]
Thanks to Ariel.
I didn't realize that the status is for the partial resident texture. That is a very important feature that many of triple A games make use of.
We definitely should consider supporting it fully.

**Comment by jkwak-work** [comment]
I think we need a new issue for fully supporting the partial resident texture feature.
This PR is to address the missing `status` for Gather functions.

**Comment by ArielG-NV** [comment]
> I think we need a new issue for fully supporting the partial resident texture feature. 
This PR is to address the missing status for Gather functions.

[Made an issue for the rest of the operations](https://github.com/shader-slang/slang/issues/4495)

File: source/slang/hlsl.meta.slang (hunk 4 above)
Comment (ArielG-NV) [reply source/slang/hlsl.meta.slang:2309]: note: Even if we don't implement all sparse texture functions this PR, we should still remove `$(statusInit)` and use the correct SPIRV/GLSL function for `Gather` (with `status`) operations.

**Review by csyonghe — COMMENTED** [review]
If we haven't meaningfully implemented status for Vulkan, we should make it a capability error instead of silently implement it in an incorrect way, that will make it hard for users to identify the problem.

File: source/slang/hlsl.meta.slang (hunk 4 above)
Comment (jkwak-work) [reply source/slang/hlsl.meta.slang:2309]: Assigning `status` to zero is how the other functions are handling it.
https://github.com/shader-slang/slang/blob/323fdff4544ac3b640bdc617d7f7d64ab56fbf5a/source/slang/hlsl.meta.slang#L705

Removing `$(statusInit)` is easy when it is the actual implemented and tested properly.

File: tests/hlsl-intrinsic/texture/texture-intrinsics.slang [hunk 5]
```diff
@@ -244,6 +246,28 @@ void computeMain(int3 dispatchThreadID: SV_DispatchThreadID)
...
+#if !defined(VK)
+    f4 = t2D.GatherGreen(samplerState, float2(u, u), int2(1, 2), status);
+    val += f4.x;  val += f4.y;  val += f4.z;  val += f4.w;
+
+    f4 = t2D.GatherGreen(samplerState, float2(u, u), int2(1, 2), int2(3, 4), int2(5, 6), int2(7, 8), status);
+    val += f4.x;  val += f4.y;  val += f4.z;  val += f4.w;
+
+    f4 = tCubeArray.GatherGreen(samplerState, float4(1.5, 1.5, 1.5, 1.5), status);
+    val += f4.x;  val += f4.y;  val += f4.z;  val += f4.w;
+#endif
```
Comment (jkwak-work) [inline tests/hlsl-intrinsic/texture/texture-intrinsics.slang:269]: @ArielG-NV , I think there is a bug on the capability check.
When `#if !defined(VK)` is not there, VK test ran just fine, which should have triggered a capability error.
Can you take a look at this?

**Comment by jkwak-work** [comment]
> If we haven't meaningfully implemented status for Vulkan, we should make it a capability error instead of silently implement it in an incorrect way, that will make it hard for users to identify the problem.

I pushed a change to apply the capability check so that only HLSL will be allowed to use the status variant.
However, it seems that there is a bug that VK can still run the function with no problem.
I asked @ArielG-NV to take a look.

**Comment by ArielG-NV** [comment]
> I pushed a change to apply the capability check so that only HLSL will be allowed to use the status variant.
However, it seems that there is a bug that VK can still run the function with no problem.
I asked @ArielG-NV to take a look.

I looked into it, there are 2 issues causing this bug:
1. `texture_gather` has incorrect capabilities (missing `hlsl`)
2. `validateAttribute` of `RequireCapabilityAttribute` never checks if the capability produced is invalid or not. This should be an error.

What happens in this scenario is:  `texture_gather` is missing `hlsl`, so `[require(hlsl, texure_gather)]` resolves into `Invalid`, which is not supposed to be possible, causing problems.

**Comment by ArielG-NV** [comment]
Seems there is a third issue: `__target_switch` never compares capabilities to the parent `[require(...)]`

File: tests/hlsl-intrinsic/texture/texture-intrinsics.slang [hunk 6]
```diff
@@ -1,7 +1,7 @@
 //DISABLE_TEST(compute):COMPARE_COMPUTE_EX:-cpu -compute -shaderobj -output-using-type
-//TEST(compute):COMPARE_COMPUTE_EX(filecheck-buffer=DX11):-slang -compute -shaderobj -output-using-type -render-feature hardware-device
-//TEST(compute):COMPARE_COMPUTE_EX(filecheck-buffer=DX12):-slang -compute -dx12 -shaderobj -output-using-type
-//TEST(compute):COMPARE_COMPUTE_EX(filecheck-buffer=DX12CS6):-slang -compute -dx12 -profile cs_6_0 -use-dxil -shaderobj -output-using-type -xslang -DCS60
+//T-EST(compute):COMPARE_COMPUTE_EX(filecheck-buffer=DX11):-slang -compute -shaderobj -output-using-type -render-feature hardware-device
```
Comment (csyonghe) [inline tests/hlsl-intrinsic/texture/texture-intrinsics.slang:2]: Why are we disabling these tests?

Comment (jkwak-work) [reply tests/hlsl-intrinsic/texture/texture-intrinsics.slang:2]: Oops. That is a mistake. I will revert them.

**Comment by jkwak-work** [comment]
> @jkwak-work do you want the fix to be in a separate PR or apart of your `jkwak-work:feature/support_GatherAlpha_with_status_argument` branch (I can make a PR to your branch)?

I think your PR #4510 will have the fix, right?
I will wait for your PR to be merged first and make sure CI goes all green with my PR.
Thanks for fixing the issue so promptly.

**Comment by jkwak-work** [comment]
With the ToT rebased, this PR is failing with capability related errors.
I will need to fix them.

**Comment by jkwak-work** [comment]
All of the issues are resolved for this PR.
It is ready for another review.

**Review by csyonghe — APPROVED** [review]

**Pull request merged**